### PR TITLE
web: initial Datadog RUM integration

### PR DESCRIPTION
**Pull request by valerybugakov**
## Context

This PR integrates [Datadog RUM](https://docs.datadoghq.com/real_user_monitoring/browser/) SDK into the web application.

Closes https://github.com/sourcegraph/sourcegraph/issues/33992
Related to: https://github.com/sourcegraph/sourcegraph/issues/29651 and [RFC 575 APPROVED: Cloud Observability tooling](https://docs.google.com/document/d/1xnAgloZB8sEkyhecjml2ByQl-aUCrJdWDYOBj3asA9g/edit#heading=h.trqab8y0kufp)

## Changes

- Datadog RUM SDK is initialized via async script along with Sentry SDK.
- Datadog RUM configuration is added to the site-config under the `observability.logging` key.
- `@datadog/browser-rum-slim` is added to `devDependencies` to allow usage of Typescript types, and it's not added to the application bundle.

### Notes integration trade-offs

Since we didn't decide which solution we would want to use for the observability OKR set for the Frontend Platform team next quarter, it's essential to integrate the Datadog RUM SDK as a POC without hurting our performance metrics on production.

- [The bundle size](https://bundlephobia.com/package/@datadog/browser-rum@4.7.1) of the default Datadog browser package [is pretty big](https://github.com/DataDog/browser-sdk/issues/317).
- They have [a slim version](https://bundlephobia.com/package/@datadog/browser-rum-slim@4.7.1) that doesn't include [the SessionReplay feature](https://docs.datadoghq.com/real_user_monitoring/session_replay/#overview), but it's still significant.
- The SDK [doesn't provide a lazy-loading out of the box](https://github.com/DataDog/browser-sdk/issues/985) without the possibility of missing some events.

The best strategy for POC is to load it [via async script](https://github.com/DataDog/browser-sdk/blob/main/packages/rum/README.md#cdn-async) to avoid hurting our bundle size. This way, we keep Sentry in place and see how they compare.

### Production

Merging this PR won't enable Datadog RUM on any production instance automatically. To enable it, the Datadog RUM configuration should be added to the site configuration via site-admin.

### Local development

Datadog is initialized only if:
1. The SDK script is included in the `index.html` template (app.html). Meaning that `SourcegraphDotComMode == true`.
2. Datadog RUM is configured using Sourcegraph site configuration.
3. `process.env.ENABLE_MONITORING || process.env.NODE_ENV === 'production'` to prevent log spam in the development environment.

Based on the requirements above to test Datadog RUM locally run: `ENABLE_MONITORING=true sg start dotcom`

### Datadog access

Datadog RUM application logs are available [here](https://app.datadoghq.com/rum/list). If the link is inaccessible for you, check if Datadog is added to your Okta profile. If it's not there — reach out to [#it-tech-ops](https://sourcegraph.slack.com/archives/C01CSS3TC75)

### Testing

1. Run `ENABLE_MONITORING=true sg start dotcom`
2. Add Dagadog configuration to the site config.
3. Check out the network tab:
    - The Datadog SDK should be loaded.
    - Pageviews should be logged to Datadog servers.

### CODE REVIEW

File: client/web/src/monitoring/initMonitoring.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+window.addEventListener('error', error => {
+    /**
+     * The "ResizeObserver loop limit exceeded" error means that `ResizeObserver` was not
+     * able to deliver all observations within a single animation frame. It doesn't break
+     * the functionality of the application. The W3C considers converting this error to a warning:
+     * https://github.com/w3c/csswg-drafts/issues/5023
+     * We can safely ignore it in the production environment to avoid hammering Sentry and other
+     * libraries relying on `window.addEventListener('error', callback)`.
+     */
+    const isResizeObserverLoopError = error.message === 'ResizeObserver loop limit exceeded'
```
Comment (jasongornall) [inline client/web/src/monitoring/initMonitoring.ts:13]: usually safari has good repro on this

File: client/web/src/monitoring/datadog/datadogClient.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    // Used to ignore some errors in the `beforeSend` hook.
+    originalException: unknown
+    errorInfo?: ErrorInfo
+}
+
+export const DatadogClient = {
+    addError: (error: unknown, context: ErrorContext): void => {
+        // Temporary solution for checking the availability of the
+        // Datadog SDK until we decide to move forward with this service.
+        if (typeof DD_RUM !== 'undefined') {
```
Comment (jasongornall) [inline client/web/src/monitoring/datadog/datadogClient.ts:15]: I'd suggest making this a helper function

File: package.json [hunk 3]
```diff
@@ -360,6 +360,7 @@
     "@codemirror/text": "^0.19.6",
     "@codemirror/tooltip": "^0.19.16",
     "@codemirror/view": "^0.19.47",
+    "@datadog/browser-rum-slim": "^4.7.1",
```
Comment (jasongornall) [inline package.json:363]: as discussed potentially move to dev dependencies

File: cmd/frontend/internal/app/ui/app.html [hunk 4]
```diff
@@ -40,7 +40,15 @@
 	'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
 	})(window,document,'script','dataLayer','GTM-TB4NLS7');</script>
 	<!-- End Google Tag Manager -->
+	<!-- Sentry -->
 	<script src='https://js.sentry-cdn.com/ae2f74442b154faf90b5ff0f7cd1c618.min.js' crossorigin="anonymous"></script>
```
Comment (jasongornall) [inline cmd/frontend/internal/app/ui/app.html:44]: woah we load sentry blocking style?

File: client/web/src/sentry/init.ts [hunk 5]
```diff
@@ -1,68 +0,0 @@
...
-window.addEventListener('error', error => {
-    /**
-     * The "ResizeObserver loop limit exceeded" error means that `ResizeObserver` was not
-     * able to deliver all observations within a single animation frame. It doesn't break
-     * the functionality of the application. The W3C considers converting this error to a warning:
-     * https://github.com/w3c/csswg-drafts/issues/5023
-     * We can safely ignore it in the production environment to avoid hammering Sentry and other
-     * libraries relying on `window.addEventListener('error', callback)`.
-     */
-    const isResizeObserverLoopError = error.message === 'ResizeObserver loop limit exceeded'
```
Comment (jasongornall) [inline client/web/src/sentry/init.ts:17]: fyi: not sure this actually works.. I see this all over sentry
https://sentry.io/organizations/sourcegraph/issues/2826145180/?environment=production&project=1391511&query=is%3Aunresolved+resize&statsPeriod=14d

File: client/web/src/monitoring/datadog/initDatadog.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+import { authenticatedUser } from '../../auth'
+import { shouldErrorBeReported } from '../shouldErrorBeReported'
+
+declare global {
+    const DD_RUM: RumGlobal
+}
+
+/**
+ * Datadog is initialized only if:
```
Comment (jasongornall) [inline client/web/src/monitoring/datadog/initDatadog.ts:12]: loving these comments!

File: cmd/frontend/internal/app/ui/app.html (hunk 4 above)
Comment (valerybugakov) [reply cmd/frontend/internal/app/ui/app.html:44]: It's a shim for lazy-loading the actual bundle.
https://github.com/sourcegraph/sourcegraph/blob/7f27382f25e0a9548520eb701de5684da1928658/client/web/src/sentry/init.ts#L34-L35

File: client/web/src/sentry/init.ts (hunk 5 above)
Comment (valerybugakov) [reply client/web/src/sentry/init.ts:17]: Yeah, I also saw this in Sentry but didn't have time to follow up. 
Before this `isResizeObserverLoopError` check, we reported a much higher volume of such errors.

**Review by eseliger — APPROVED** [review]

File: client/web/src/jscontext.ts [hunk 7]
```diff
@@ -24,6 +24,12 @@ export interface SourcegraphContext extends Pick<Required<SiteConfiguration>, 'e
 
     readonly sentryDSN: string | null
 
+    /* Configuration required for Datadog RUM (https://docs.datadoghq.com/real_user_monitoring/browser/#setup). */
```
Comment (umpox) [inline client/web/src/jscontext.ts:27]:
```suggestion
    /** Configuration required for Datadog RUM (https://docs.datadoghq.com/real_user_monitoring/browser/#setup). */
```

Needs this to be valid JSDoc and work for hover definitions I think :)

File: client/web/src/monitoring/datadog/datadogClient.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import { ErrorInfo } from 'react'
+
+// Ensures consistent `ErrorContext` across `addError` calls.
+// https://docs.datadoghq.com/real_user_monitoring/browser/modifying_data_and_context/?tab=npm#enrich-and-control-rum-data
```
Comment (umpox) [inline client/web/src/monitoring/datadog/datadogClient.ts:4]:
```suggestion
/**
 * Ensures consistent `ErrorContext` across `addError` calls.
 * https://docs.datadoghq.com/real_user_monitoring/browser/modifying_data_and_context/?tab=npm#enrich-and-control-rum-data
 */
```

For hover definitions, especially as we export a func that uses this

**Review by umpox — APPROVED** [review]

**Pull request merged**